### PR TITLE
Remove redundant dirty updates

### PR DESCRIPTION
**Pull request by kchibisov**
In some cases dirty was set without any ui update leading
to extra redraws, this commit resolves this.

Co-authored-by: Greg Depoire--Ferrer <greg@gregdf.com>

cc @greg904

### CODE REVIEW

File: alacritty/src/event.rs [hunk 1]
```diff
@@ -286,9 +287,8 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
 
     fn start_selection(&mut self, ty: SelectionType, point: Point, side: Side) {
         self.terminal.selection = Some(Selection::new(ty, point, side));
-        *self.dirty = true;
-
         self.copy_selection(ClipboardType::Selection);
+        *self.dirty = true;
```
Comment (chrisduerr) [inline alacritty/src/event.rs:291]: Why was this moved? I actually kinda like putting it under the statement that actually requires the dirty flag.

The `copy_selection` doesn't care about dirty or not, right? So why not have the statement closer to what actually caused the dirty change?

File: alacritty/src/event.rs [hunk 2]
```diff
@@ -651,13 +647,13 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
         let timer_id = TimerId::new(Topic::BlinkCursor, self.display.window.id());
         if self.scheduler.unschedule(timer_id).is_some() {
             self.schedule_blinking();
-            self.display.cursor_hidden = false;
+            if mem::replace(&mut self.display.cursor_hidden, false) {
+                *self.dirty = true;
+            }
```
Comment (chrisduerr) [inline alacritty/src/event.rs:652]: I'd add an empty line above and put a quick comment that we only need to redraw if the cursor is currently hidden.

Also since this is a `bool` you can use `mem::take` instead.

File: alacritty/src/event.rs [hunk 3]
```diff
@@ -716,6 +712,8 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
                 self.terminal.vi_goto_point(*hint_bounds.start());
             },
         }
+
+        self.mark_dirty();
```
Comment (chrisduerr) [inline alacritty/src/event.rs:716]: Which action here requires being marked dirty? We shouldn't just randomly sprinkle this around because it seems convenient. Most of the actions taken here don't seem to require any damage?

The only things I could imagine would be `MoveViModeCursor` and `Select`?

File: alacritty/src/event.rs [hunk 4]
```diff
@@ -930,7 +928,6 @@ impl<'a, N: Notify + 'a, T: EventListener> ActionContext<'a, N, T> {
     fn exit_search(&mut self) {
         self.display.pending_update.dirty = true;
         self.search_state.history_index = None;
-        *self.dirty = true;
```
Comment (chrisduerr) [inline alacritty/src/event.rs:933]: This seems **really** wrong. There definitely should be some dirtiness going on when exiting the search. At the very least when `focused_match != None`.

File: alacritty/src/event.rs [hunk 5]
```diff
@@ -1084,14 +1080,13 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                     // Disable blinking after timeout reached.
                     let timer_id = TimerId::new(Topic::BlinkCursor, self.ctx.display.window.id());
                     self.ctx.scheduler.unschedule(timer_id);
-                    self.ctx.display.cursor_hidden = false;
                     *self.ctx.cursor_blink_timed_out = true;
+                    self.ctx.display.cursor_hidden = false;
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1084]: Nice, looks much more pretty this way.

File: alacritty/src/event.rs [hunk 6]
```diff
@@ -1184,7 +1177,6 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                     },
                     WindowEvent::Focused(is_focused) => {
                         self.ctx.terminal.is_focused = is_focused;
-                        *self.ctx.dirty = true;
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1187]: Removing this without cursor blinking might cause issues? We **must** have a dirty flag set whenever focus changed because of the box cursor change.

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/event.rs [hunk 7]
```diff
@@ -236,7 +237,8 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
             self.update_selection(point, self.mouse.cell_side);
         }
 
-        *self.dirty = true;
+        // Update dirty if actually scrolled or we're in the Vi mode.
+        *self.dirty |= lines_changed != 0 || vi_mode;
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty/src/event.rs:241]: Since scroll can move Vi cursor.

File: alacritty/src/event.rs (hunk 1 above)
Comment (kchibisov) [reply alacritty/src/event.rs:291]: I've sorted them by length.

File: alacritty/src/event.rs (hunk 3 above)
Comment (kchibisov) [reply alacritty/src/event.rs:716]: Yes, due to Vi movements it's here.

File: alacritty/src/event.rs (hunk 4 above)
Comment (kchibisov) [reply alacritty/src/event.rs:933]: There's dirty already `self.display.pending_update.dirty = true;`. No need to set it twice.

File: alacritty/src/event.rs (hunk 6 above)
Comment (kchibisov) [reply alacritty/src/event.rs:1187]: Oh, forgot about it, will update.

File: alacritty/src/event.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisduerr) [reply alacritty/src/event.rs:241]: Even if the display doesn't move at all?

File: alacritty/src/event.rs (hunk 1 above)
Comment (chrisduerr) [reply alacritty/src/event.rs:291]: Yeah I'm not sure that's the best choice here, especially since there's a newline in between?

File: alacritty/src/event.rs (hunk 3 above)
Comment (chrisduerr) [reply alacritty/src/event.rs:716]: Should only be in the vi branches then probably?

File: alacritty/src/event.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty/src/event.rs:241]: Yes, we clamp it to viewport boundaries from what I can see.

```rust
// Clamp vi mode cursor to the viewport.
let viewport_start = -(self.grid.display_offset() as i32);
let viewport_end = viewport_start + self.bottommost_line().0;
let vi_cursor_line = &mut self.vi_mode_cursor.point.line.0;
*vi_cursor_line = cmp::min(viewport_end, cmp::max(viewport_start, *vi_cursor_line));
```

[the remark this answers is not in the document]
Comment (chrisduerr) [reply alacritty/src/event.rs:241]: Yeah but that shouldn't do anything, right? Like the vi cursor should always be clamped, scrolling isn't meant just for clamping?

I thought the clamping here was only necessary if we actually do scroll?

[the remark this answers is not in the document]
Comment (kchibisov) [reply alacritty/src/event.rs:241]: Hm, seems like you are right here. We don't need an extra check for vi mod scroll.

File: alacritty/src/event.rs [hunk 8]
```diff
@@ -714,6 +716,7 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
                 }
 
                 self.terminal.vi_goto_point(*hint_bounds.start());
+                self.mark_dirty();
```
Comment (chrisduerr) [inline alacritty/src/event.rs:719]: Do we also need to `mark_dirty` in the `HintInternalAction::Select`, or is that handled by `update_selection`? I'd assume the latter?

File: alacritty/src/event.rs [hunk 9]
```diff
@@ -1184,7 +1182,11 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                     },
                     WindowEvent::Focused(is_focused) => {
                         self.ctx.terminal.is_focused = is_focused;
-                        *self.ctx.dirty = true;
+
+                        // When the unfocused hollow is used we must redraw on focus change.
+                        if self.ctx.config.terminal_config.cursor.unfocused_hollow {
+                            *self.ctx.dirty = false;
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1188]:
```suggestion
                            *self.ctx.dirty = true;
```

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/event.rs (hunk 8 above)
Comment (kchibisov) [reply alacritty/src/event.rs:719]: Yes, update selection does that.

**Review by chrisduerr — APPROVED** [review]
